components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldLayoutTests.m

```diff
@@ -0,0 +1,37 @@
+// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <XCTest/XCTest.h>
+
+#import "MDCBaseTextFieldLayout.h"
+#import "MaterialTextFields+ContainedInputView.h"
+
+@interface MDCBaseTextFieldLayout (Testing)
+- (CGFloat)textHeightWithFont:(UIFont *)font;
+@end
+
+@interface MDCBaseTextFieldLayoutTests : XCTestCase
+@end
+
+@implementation MDCBaseTextFieldLayoutTests
+
+- (void)testTextHeightWithFont {
+  MDCBaseTextFieldLayout *emptyLayout = [[MDCBaseTextFieldLayout alloc] init];
+  UIFont *systemFont = [UIFont systemFontOfSize:[UIFont systemFontSize]];
+  CGFloat ceiledLineHeight = (CGFloat)ceil((double)systemFont.lineHeight);
+  CGFloat calculatedTextHeight = [emptyLayout textHeightWithFont:systemFont];
+  XCTAssertEqual(ceiledLineHeight, calculatedTextHeight);
+}
```

Comment:
Please test the full set of functionality you've added thus far. Tests will help ensure that we're not regressing on any of the edge cases you've been calling out throughout the code documentation, and it'll be much easier to start writing the tests now than to do them all at once later.

At minimum there needs to be tests for:

- `textRectForBounds:` and the edge case behavior you documented. Specifically write a test for the case you described in the comments so that we don't accidentally regress on it.
- Each of the different configurations currently possible in the layout object (e.g. only a left view, only a right view, both views, no views).
- Each of the different combinations of display modes for both the left and right view.

Reply:
I added a bunch of tests for the leading/trailing view stuff.

`-textRectForBounds:` just calls `-adjustTextAreaFrame:withParentClassTextAreaFrame:`, which now has a unit test. Does `-textRectForBounds:` really need to be tested in light of that? Also, I wouldn't call the text rect behavior an edge case, it's just how the UITextField text rect stuff works, in every case.

Reply:
Yes, please test all of the public APIs that you're customizing to ensure that the wiring you've implemented here does not regress.